File: backend/CertificatesManagerApi/Mappers/CertificateMapper.cs
using CertificatesManagerApi.DTOs;
using CertificatesManagerApi.Entities;
using CertificatesManagerApi.Utils;

namespace CertificatesManagerApi.Mappers
{
    public class CertificateMapper
    {
        public static CertificateDTO ToDto(Certificate certificate)
        {
            CertificateDTO certificateDto = new CertificateDTO
            {
                Handle = certificate.Handle,
                Supplier = SupplierMapper.ToDto(certificate.Supplier),
                CertificateType = CertificateTypeMapper.ToDto(certificate.Type),
                ValidFrom = DateConvert.DateToString(certificate.ValidFrom),
                ValidTo = DateConvert.DateToString(certificate.ValidTo),
                Document = certificate.CertificateDocument,
                Comments = certificate.Comments
                .Select(comment => CommentMapper.ToDto(comment)).ToList(),
                Participants = certificate.CertificateUsers
                .Select(
                    certificateUser => UserMapper.ToDto(certificateUser.User)
                )
                .ToList()
            };
            return certificateDto;
        }

        public static TableCertificatesDTO ToMultipleDto(Certificate certificate)
        {
            TableCertificatesDTO certificatesDto = new TableCertificatesDTO
            {
                Handle = certificate.Handle,
                Supplier = @$"{certificate.Supplier.Name}, {certificate.Supplier.Index}, {certificate.Supplier.City}",
                CertificateType = certificate.Type.Type,
                ValidFrom = DateConvert.DateToString(certificate.ValidFrom),
                ValidTo = DateConvert.DateToString(certificate.ValidTo)
            };
            return certificatesDto;
        }

        public static Certificate ToEntity(CreateCertificateDTO certificateDTO)
        {
            Certificate certificate = new()
            {
                ValidFrom = DateConvert.StringToDate(certificateDTO.ValidFrom),
                ValidTo = DateConvert.StringToDate(certificateDTO.ValidTo),
                CertificateDocument = certificateDTO.Document
            };

            return certificate;
        }

        public static Certificate ToUpdateEntity(UpdateCertificateDTO certificateDTO)
        {

            Certificate certificate = new()
            {
                ValidFrom = DateConvert.StringToDate(certificateDTO.ValidFrom),
                ValidTo = DateConvert.StringToDate(certificateDTO.ValidTo),
                CertificateDocument = certificateDTO.Document
            };

            return certificate;
        }
    }
}

File: frontend/src/common/utils/mappers/certificate-mapper.ts
import Certificate from '../../models/certificate.model';
import {
	CertificateAddDTO,
	CertificateDTO,
	CertificateUpdateDTO,
} from '../../models/dtos/certificate-dto';
import { toIsoString } from '../format-date.utils';
import { CertificateTypeMapper } from './certificate-type-mapper';
import { CommentMapper } from './comment-mapper';
import { SupplierMapper } from './supplier-mapper';
import { UserMapper } from './user-mapper';

export class CertificateMapper {
	static ToModel(certificateDTO: CertificateDTO): Certificate {
		let certificate: Certificate = {
			id: certificateDTO.handle,
			supplier: SupplierMapper.ToModel(certificateDTO.supplier),
			type: CertificateTypeMapper.ToModel(certificateDTO.certificateType),
			validFrom: new Date(certificateDTO.validFrom),
			validTo: new Date(certificateDTO.validTo),
			users: certificateDTO.participants
				? certificateDTO.participants.map((user) => UserMapper.ToModel(user))
				: [],
			pdf: certificateDTO.document,
			comments: certificateDTO.comments
				? certificateDTO.comments.map((comment) =>
						CommentMapper.ToModel(comment),
					)
				: [],
		};

		return certificate;
	}

	static ToAddDTO(certificate: Certificate): CertificateAddDTO {
		let certificateAddDTO: CertificateAddDTO = {
			supplierHandle: certificate.supplier.id,
			certificateTypeHandle: certificate.type.id,
			validFrom: toIsoString(certificate.validFrom),
			validTo: toIsoString(certificate.validTo),
			document: certificate.pdf,
		};

		return certificateAddDTO;
	}

	static ToUpdateDTO(certificate: Certificate): CertificateUpdateDTO {
		let certificateUpdateDTO: CertificateUpdateDTO = {
			supplierHandle: certificate.supplier.id,
			certificateTypeHandle: certificate.type.id,
			validFrom: toIsoString(certificate.validFrom),
			validTo: toIsoString(certificate.validTo),
			document: certificate.pdf,
			participants: certificate.users?.map((user) => UserMapper.ToDTO(user)),
		};

		return certificateUpdateDTO;
	}
}
